### PR TITLE
[mono] fix SIMD instructions' availability

### LINKED ISSUE #34304: Review `MONO_CPU_X86_SSSE3_COMBINED`
`MONO_CPU_X86_SSSE3_COMBINED` is currently defined as `MONO_CPU_X86_SSE3_COMBINED  | MONO_CPU_X86_PCLMUL | MONO_CPU_X86_AES | MONO_CPU_X86_SSSE3,` (https://github.com/dotnet/runtime/blob/335ffb05fbe97187452046bc6791405ea33f210c/src/mono/mono/mini/mini.h#L2832).

However, this doesn't match the hierarchy exposed by `System.Runtime.Intrinsics` nor does it match the CPUID checks specified by the Intel or AMD architecture manuals:

From `Intel® 64 and IA-32 Architectures Software Developer’s Manual Volume 1: Basic Architecture, October 2019`
image
image

### PR DESCRIPTION
**Pull request by EgorBo**
Fixes: https://github.com/dotnet/runtime/issues/34304

@tannergooding @davidwrighton 
Now it fully matches clang/LLVM behavior, e.g. even `-mattr=+avx2` (or `-mavx2` in clang) doesn't enable `pcmul` and `aes`. However, `-mattr=+sse4.2` (`-msse4.2`) enables `popcnt` (see https://godbolt.org/z/RWAp_D). The intel manual says

image

So is there a CPU with AVX set but without `popcnt`? The instruction is pretty useful (it is used in BCL) users might forget to define it when they use `mattr=avx` (or `mattr=avx2`).

However in C++ world programmers use named groups e.g. `mcpu=haswell` rather than listing these sets via attributes.

### CODE REVIEW

**Review by vargaz — APPROVED** [review]

**Comment by tannergooding** [comment]
> So is there a CPU with AVX set but without popcnt? 

I would expect that this wouldn't happen in practice, but I think it is still appropriate to model the underlying checks accurately so we aren't stuck with a breaking change in the future if that ever changes.

For Intel, `POPCNT` was introduced as part of `SSE4.2` (2008, Nehalem), but it is a separate CPUID check because AMD had already introduced `POPCNT` and `LZCNT` independently as part of ABM (2007, Barcelona). Intel didn't introduce `LZCNT` until much later alongside AVX2/BMI1/BMI2/FMA3 (2013, Haswell).

File: src/mono/mono/mini/mini.h [hunk 1]
```diff
@@ -2829,14 +2829,15 @@ typedef enum {
...
 	MONO_CPU_X86_SSE41_COMBINED       = MONO_CPU_X86_SSSE3_COMBINED | MONO_CPU_X86_SSE41,
 	MONO_CPU_X86_SSE42_COMBINED       = MONO_CPU_X86_SSE41_COMBINED | MONO_CPU_X86_SSE42,
 	MONO_CPU_X86_POPCNT_COMBINED      = MONO_CPU_X86_SSE42_COMBINED | MONO_CPU_X86_POPCNT,
 	MONO_CPU_X86_AVX_COMBINED         = MONO_CPU_X86_SSE42_COMBINED | MONO_CPU_X86_AVX,
-	MONO_CPU_X86_AVX2_COMBINED        = MONO_CPU_X86_AVX_COMBINED   | MONO_CPU_X86_POPCNT | MONO_CPU_X86_AVX2,
-	MONO_CPU_X86_FMA_COMBINED         = MONO_CPU_X86_AVX_COMBINED   | MONO_CPU_X86_POPCNT | MONO_CPU_X86_FMA,
-	MONO_CPU_X86_FULL_SSEAVX_COMBINED = MONO_CPU_X86_FMA_COMBINED   | MONO_CPU_X86_AVX2,
+	MONO_CPU_X86_AVX2_COMBINED        = MONO_CPU_X86_AVX_COMBINED   | MONO_CPU_X86_AVX2,
+	MONO_CPU_X86_FMA_COMBINED         = MONO_CPU_X86_AVX_COMBINED   | MONO_CPU_X86_FMA,
+	MONO_CPU_X86_FULL_SSEAVX_COMBINED = MONO_CPU_X86_FMA_COMBINED   | MONO_CPU_X86_AVX2   | MONO_CPU_X86_PCLMUL
```
Comment (tannergooding) [inline src/mono/mono/mini/mini.h]: This name is a bit confusing, I'm not quite sure based on the name along, what is meant to be enabled by it.

**Review by tannergooding — APPROVED** [review]
LGTM and to match the hierarchy specified by the Intel/AMD architecture manuals.

File: src/mono/mono/mini/mini.h (hunk 1 above)
Comment (EgorBo) [reply src/mono/mono/mini/mini.h]: @tannergooding it's only used internally to be able to disable SSE/AVX like that:
`mattr=-sse3` => disable sse3, also ssse3, sse4.1, etc (but don't touch lzcnt, bmit)

see here: https://github.com/dotnet/runtime/blob/master/src/mono/mono/mini/aot-compiler.c#L8109-L8111

it also matches llvm behaviour.

Comment (EgorBo) [reply src/mono/mono/mini/mini.h]: it allows to do things like this `mcpu=native,mattr=-avx`

**Comment by EgorBo** [comment]
The failing job is known and unrelated

**Pull request merged**